pkg/vere/main.c

```diff
@@ -1215,29 +1226,60 @@ _cw_eval(c3_i argc, c3_c* argv[])
       exit(1);
     }
   }
-
-  printf("eval:\n");
-
-  //  +wish for an eval gate (virtualized twice for pretty-printing)
-  //
-  u3_noun gat = u3v_wish("|=(a=@t (sell (slap !>(+>.$) (rain /eval a))))");
-  u3_noun res;
-  {
+  fprintf(stderr, "eval:\n");
+  if ( c3n == jam_o ) {
+    //  +wish for an eval gate (virtualized twice for pretty-printing)
+    //
+    u3_noun gat = u3v_wish("|=(a=@t (sell (slap !>(+>.$) (rain /eval a))))");
+    u3_noun res;
+    {
+      u3_noun sam = u3i_string(evl_c);
+      u3_noun cor = u3nc(u3k(u3h(gat)), u3nc(sam, u3k(u3t(u3t(gat)))));
+      res = u3m_soft(0, u3n_kick_on, cor);
+    }
+    if ( 0 == u3h(res) ) {  //  successful execution, print output
+      u3_pier_tank(0, 0, u3k(u3t(res)));
+    } else {                  //  error, print stack trace
+       u3_pier_punt_goof("eval", u3k(res));
+    }
+    u3z(res);
+    u3z(gat);
+  } else {
     u3_noun sam = u3i_string(evl_c);
-    u3_noun cor = u3nc(u3k(u3h(gat)), u3nc(sam, u3k(u3t(u3t(gat)))));
-    res = u3m_soft(0, u3n_kick_on, cor);
-  }
-
-
-  if ( 0 == u3h(res) ) {  //  successful execution, print output
-     u3_pier_tank(0, 0, u3k(u3t(res)));
-  }
-  else {                  //  error, print stack trace
-     u3_pier_punt_goof("eval", u3k(res));
+    u3_noun res = u3v_wish_n(sam);
```

Comment:
This evaluation should be virtualized with `u3m_soft()`. This is the reason to add `u3v_wish_n()` instead of just calling `u3v_wish()`.

The implementations of those two should also be combined at some point (and arguably the names should be reversed: `u3v_wish()` and `u3v_wish_c()` would be more consistent with u3 conventions). But that can be done some other time -- it would complicate this PR.

Reply:
When I try to replicate the `u3m_soft()` I do for the non jam case the result is incorrect. I think its because I'm jamming a vase but I'm not sure.

Reply:
`u3m_soft(0, u3v_wish_n, sam)`. Then, just as in the previous case, you'll need to branch on success/failure (`0 == u3h(res)`) and print the trace or proceed appropriately.

Reply:
Whenever I do that I get
```
$ echo "(add 2 2)" | bazel-bin/pkg/vere/urbit eval --jam
loom: mapped 2048MB
lite: arvo formula 2a2274c9
lite: core 4bb376f0
lite: final state 4bb376f0
eval:
jammed noun: 
home: bailing out
Aborted (core dumped)
```

The code change is below:
```
else {
     u3_noun sam = u3i_string(evl_c);
     u3_noun res = u3m_soft(0, u3v_wish_n, sam);
     c3_d bits = 0;
     c3_d len_d = 0;
     c3_y* byt_y;
     if ( 0 == u3h(res) ) {  //  successful execution, print output
        bits = u3s_jam_xeno(res, &len_d, &byt_y);
        if ( c3n == kan_o ) {
          fprintf(stderr,"jammed noun: ");
          for ( size_t p=0; p < len_d; p++ ){
            fprintf(stdout,"\\x%2x", byt_y[p++]);
          } 
          fprintf(stderr,"\n");
 ```
  ...
``` 
     } else {                  //  error, print stack trace
        u3_pier_punt_goof("eval", u3k(res));
     }  
```

Reply:
You'll want to `u3s_jam_xeno(u3t(res) ...)`. But I don't see why anything should crash here. Can you reproduce that in a debugger with a breakpoint on `u3m_bail()`?

Reply:
```
Breakpoint 1 at 0x43bdf0: file pkg/noun/manage.c, line 688.
(gdb) run eval --jam < ~/add.hoon
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/amadeo/.cache/bazel/_bazel_amadeo/bfa2b923decdeb8d329ce362a61e8ce7/execroot/__main__/bazel-out/k8-fastbuild/bin/pkg/vere/urbit eval --jam < ~/add.hoon
loom: mapped 2048MB
lite: arvo formula 2a2274c9
lite: core 4bb376f0
lite: final state 4bb376f0
eval:
jammed noun: 

Breakpoint 1, u3m_bail (how=how@entry=1819635558) at pkg/noun/manage.c:688
688	  if ( &(u3H->rod_u) == u3R ) {
(gdb) bt
#0  u3m_bail (how=how@entry=1819635558) at pkg/noun/manage.c:688
#1  0x0000000000428414 in _me_lose_north (dog=2150035627) at pkg/noun/allocate.c:1375
#2  0x0000000000402d83 in _cw_eval (argv=<optimized out>, argc=<optimized out>)
    at pkg/vere/main.c:1286
#3  _cw_utils (argv=<optimized out>, argc=<optimized out>) at pkg/vere/main.c:2029
#4  main (argc=<optimized out>, argv=<optimized out>) at pkg/vere/main.c:2064
(gdb) 
```

The relevent part of the eval loop is below:
```
1281     } else {                  //  error, print stack trace
1282        u3_pier_punt_goof("eval", u3k(res));
1283     }
1284 
1285     u3z(res);
1286     u3z(sam);
1287   }
1288   free(evl_c);
1289 }
```

I think I'm freeing `sam` when I dont need to because I'm using it in `u3m_soft()`. I removed line 1286 and it seems to stop crashing

Reply:
Pushed up the changes

Reply:
Good catch!